### ISSUE TITLE
ansi color code highlight util

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#30415]
# Problem

Nvim is a pretty good pager for "man", but can't easily replace `less -R` for other (non-manpage) terminal output containing ANSI color codes.

Programs like kitty terminal, allow configuring any program such as `less -R` or `nvim` for displaying terminal "scrollback". `less -R` colorizes ANSI color codes, but Nvim doesn't.

Currently, this is the `kitty.config` setting I use:

    scrollback_pager nvim --clean --cmd 'set eventignore=FileType' +'%s/\e\[[0-9;]*m//g' +'set nomodified readonly' +'$' -

but that removes the escape codes instead of colorizing them.

# Expected behavior

- ✅ document `nvim_open_term` snippet
- ✅ ~~provide a `nvim_open_term` wrapper as command/function~~ to colorize a buffer containing ANSI escape codes.
    - #33720
    - https://github.com/folke/dot/blob/39602b7edc7222213bce762080d8f46352167434/nvim/lua/util/init.lua#L68-L93
    - https://github.com/neovim/neovim/issues/30415#issuecomment-2368519968
- Can handle [streamed input](https://github.com/neovim/neovim/issues/30415#issuecomment-2750312295) (and more than 100k lines)
- Expose `:terminal highlights` as highlight groups. https://github.com/neovim/neovim/pull/7406#issuecomment-337363444

## Reference

previous implementations:

- https://github.com/neovim/neovim/issues/5054 and other posts mention ye olde AnsiEsc.vim 
- https://github.com/lucc/nvimpager

**Comment by kovidgoyal** [comment]
You might also want to take a look at: https://github.com/mikesmithgh/kitty-scrollback.nvim

**Comment by lucc** [comment]
I tried to do it with `:term` a long while ago but was not successful. My experiments are documented here: https://vi.stackexchange.com/questions/16624/use-terminal-to-display-file-with-ansi-escape-codes

The two problems I had were:
1. I can not remove the `[Process exited]` text from the buffer
2. the buffer contains extra empty lines to fill the window

I re-implemented the logic of the old [AnsiEsc.vim](http://www.drchip.org/astronaut/vim/index.html#ANSIESC) plugin in nvimpager. The code is here https://github.com/lucc/nvimpager/blob/main/lua/nvimpager/ansi2highlight.lua
It is a state machine that walks over the buffer text and dynamically creates highlight groups and applies them to the relevant parts of the buffer. If the decision is to implement something like this in neovim I can send a PR or help people implement it. (AnsiEsc.vim uses regexes to define a fixed list of highlight groups up front whereas nvimpager parses the buffer with a custom state machine and only creates the required highlight groups on the fly and applies them by buffer position)


I did not investigate if treesitter can be used to parse ansi escape sequences. Mainly because the were not present in neovim when I started nvimpager.

**Comment by glepnir** [comment]
> I can not remove the [Process exited] text from the buffer

issue track https://github.com/neovim/neovim/issues/14986

**Comment by justinmk** [comment]
> If the decision is to implement something like this in neovim I can send a PR

Yes! It looks like more code than I was expecting, so it can live in its own module which won't be loaded unless something require()'s it. I assume we will expose it as a single function on `vim.highlight` , unless there are better ideas (do we need a `vim.term` module? cc @gpanders).

**Comment by gpanders** [comment]
We shouldn't need to parse escape sequences ourselves, we can use `nvim_open_term()` and copy/pipe data from stdin into the terminal channel.

Here's a rudimentary example (probably missing a lot of edge cases):

```lua
vim.api.nvim_create_user_command("Term", function(args)
    local buf = vim.api.nvim_get_current_buf()
    local b = vim.api.nvim_create_buf(false, true)
    local chan = vim.api.nvim_open_term(b, {})
    vim.api.nvim_chan_send(chan, table.concat(vim.api.nvim_buf_get_lines(buf, 0, -1, false), "\n"))
    vim.api.nvim_win_set_buf(0, b)
end, {})
```

This creates a `:Term` user command, which you can use like this:

```
git grep --color=always foo | nvim +Term
```

And now Nvim acts as a pager with correct colored output.

**Comment by justinmk** [comment]
> we can use `nvim_open_term()` and copy/pipe data from stdin into the terminal channel.

Thanks. Yes, that's an option... is there a benefit to supporting this for "normal" buffers though? Would be cool if we could extract/copy the `:terminal` result (including highlight groups) into a normal buffer.

Maybe that should be the focus instead of reinventing termcode parsing in a plugin.

**Comment by mfussenegger** [comment]
> Yes, that's an option... is there a benefit to supporting this for "normal" buffers though

Just to add as data point:
For plugins like nvim-dap it would be convenient to have that capability because they added [ANSI styling for variables to the spec] (https://github.com/microsoft/debug-adapter-protocol/pull/501).

**Comment by LunarLambda** [comment]
> > we can use `nvim_open_term()` and copy/pipe data from stdin into the terminal channel.
> 
> Thanks. Yes, that's an option... is there a benefit to supporting this for "normal" buffers though? Would be cool if we could extract/copy the `:terminal` result (including highlight groups) into a normal buffer.
> 
> Maybe that should be the focus instead of reinventing termcode parsing in a plugin.

I think supporting normal buffers could be handy for non-pipe scenarios. Sometimes for any number of reasons you can end up with ANSI escapes in a regular file (log files, or colorized viewings of things like saved grep/ip/diff/etc outputs.)

**Comment by justinmk** [comment]
It definitely makes sense to leverage Nvim's builtin PTY features instead of reinventing that in Lua.

The code snippet in https://github.com/neovim/neovim/issues/30415#issuecomment-2368519968 works nicely and perhaps we should provide it by default so that it's easy for users to use Nvim as a pager.



> supporting normal buffers could be handy for non-pipe scenarios

To support that, the next step would be for `:terminal` colors to be exposed as syntax groups. Partially implemented here: https://github.com/neovim/neovim/pull/7406#issuecomment-337363444

**Comment by phanen** [comment]
Linux only one-line solution steal from  https://github.com/rydesun/dotfiles/blob/0e1cc9ec0dc4e0df991441c434c1ecd69f187f2f/.config/kitty/kitty.conf#L24.
```sh
fd --color=always | nvim 63<&0 </dev/null -c "au VimEnter * term cat /dev/fd/63 && sleep 10000"
```

**Comment by tmccombs** [comment]
There are several limitations to the code snippet in https://github.com/neovim/neovim/issues/30415#issuecomment-2368519968

1. the terminal mode has a hard limit of 100K lines. That is usually more than enough. But it is frustrating when  you run into it (which I have, with very long log outputs). 
2. It has to load the entire file into a buffer then copy the entire thing, rather than streaming the content into the pseudo terminal in batches . I've tried to work around this but it is surprisingly difficult. 
    - There isn't a way to intercept stdin and send it to a terminal instead of a new buffer (see https://github.com/neovim/neovim/issues/9462)
    - This is again primarily an issue with large inputs
    - Loading the buffer and then changing it can cause a visual stutter, especially if the input is large
4. It creates an additional buffer, leaving behind the old one. When reading from stdin, this means that attempting to close nvim warns you that you have unsaved changes in the original buffer. This one is easy to remedy though.

**Comment by tmccombs** [comment]
Currently `nvim_open_term` expects the buffer it is given to be empty. What if that was changed so that if it is given a buffer that is non-empty it starts by processing the contents of that buffer as if it had been passed over the channel?

**Comment by gpanders** [comment]
> Currently `nvim_open_term` expects the buffer it is given to be empty. What if that was changed so that if it is given a buffer that is non-empty it starts by processing the contents of that buffer as if it had been passed over the channel?

https://github.com/neovim/neovim/pull/33720

**Comment by justinmk** [comment]
Considering this closed, though the "Expected behavior" lists 2 followup features.

**Comment by lewis6991** [comment]
One of the problems with the proposed solution of using `nvim_open_term()` is that it has a limited maximum scrollback of 100000.

Right now I'm trying to debug the terminal history of size 250000 (for a single command), which means `nvim_open_term()` isn't an acceptable solution.

My workaround is to just strip the ansi codes and use a normal buffer:

```lua
tmux capture-pane -epS - | sed -e 's/\x1b\[[0-9;]*m//g' | nvim -
```

Is there a reason for the upper limit on `scrollback`? Or is it just a number that seemed "big enough" at the time?

Raised #34880 to increase the max value of scrollback.

**Comment by justinmk** [comment]
> Is there a reason for the upper limit on `scrollback`? Or is it just a number that seemed "big enough" at the time?

yeah, no reason. could be infinite, but eventually should think about how to dynamically let the user know when they are paying a cost for a GB-sized buffer whose scrollback they might not actually care about...

**Comment by tmccombs** [comment]
I think having a scrollback limit that is set to something by default is fine. But it should be possible to set that limit higher than 100000 if desired.
